System/Console/ANSI/Color.hs

```diff
@@ -75,46 +85,85 @@ clamp n min_ max_
-                   -> Either (ColorIntensity, Color) Color8
-color8CodeToEither (Color8Code c)
-  | c < 16    = Left $ fst $ aNSIColors !! fromIntegral c   -- interpreted as a 4-bit ANSI color
-  | c < 232   = Right $ asRGB (c-16)          -- interpreted as 8-bit rgb
-  | otherwise = Right $ Gray8Color (c-232)    -- interpreted as 8-bit grayscale
+-- The colors described by the 2 first ranges (0x00-0x07 and 0x08-0x0F) match
+-- exactly with the 4 bit ANSI colors represented by (ColorIntensity, Color).
+-- The colors of the 3rd range (0x10-0xE7) are represented by RGB8Color, and the
+-- colors of the 4th range are represented by Gray8Color.
+color8CodeToXterm256 :: Color8Code -> Xterm256Color
+color8CodeToXterm256 (Color8Code c)
+  | c < 16    = SystemColor intensity color -- interpreted as a 4-bit ANSI color
+  | c < 232   = RGBColor $ asRGB (c - 16)   -- interpreted as 8-bit rgb
+  | otherwise = GrayColor (c - 232)         -- interpreted as 8-bit grayscale
  where
-  asRGB i = let -- we know that i = 36 × r + 6 × g + b and (0 ≤ r, g, b ≤ 5) (cf. comment on top)
-                -- so we can deduce the unique set of corresponding r g and b values:
+  (intensity, color) = fst $ aNSIColors !! fromIntegral c
+  asRGB i = let -- we know that i = 36 × r + 6 × g + b and (0 ≤ r, g, b ≤ 5)
+                -- (cf. comment on top) so we can deduce the unique set of
+                -- corresponding r g and b values:
                 r = quot i 36
                 g = quot (r - 36*i) 6
                 b = i - (6*g + 36*i)
-            in  RGB8Color r g b
+            in  RGB r g b
+
+-- | Color8 represents an 8-bit ANSI color. This function converts it to the
+-- corresponding code, defined by https://en.wikipedia.org/wiki/ANSI_escape_code#Colors
+--
+-- For safety the values of Color8 are clamped in their respective ranges.
+xterm256ColorToCode :: Xterm256Color -> Color8Code
+xterm256ColorToCode (SystemColor Dull color)
+  = Color8Code (fromIntegral $ colorToCode color)
+xterm256ColorToCode (SystemColor Vivid color)
+  = Color8Code (fromIntegral $ 16 + colorToCode color)
+-- 8-bit rgb colors are represented by code:
+-- 16 + 36 × r + 6 × g + b (0 ≤ r, g, b ≤ 5) (see link to spec above)
+xterm256ColorToCode (RGBColor (RGB r' g' b'))
+  = Color8Code (16 + 36 * r + 6 * g + b)
+  where
+    clamp' x = clamp x 0 5
+    r = clamp' r'
+    g = clamp' g'
+    b = clamp' b'
+-- 8-bit grayscale colors are represented by code: 232 + g (g in [0..23]) (see
+-- link to spec above)
+xterm256ColorToCode (GrayColor y) = Color8Code (232 + clamp y 0 23)
 
--- | Finds the 4 bit color that is closest in srgb colorspace to a given 24-bit color
-closest4bitsANSIColor :: SRGB.Colour Float -> (ColorIntensity, Color)
-closest4bitsANSIColor color = fst $ minimumBy order aNSIColors
+-- | Finds the color in the xterm 256 color protocol that is closest in srgb
+-- colorspace to a given color in that colorspace
+closestXterm256Color :: Colour Float -> Xterm256Color
```

Comment:
I don't see where this function is used?

Reply:
It is provided only as a utility function, exported by `System.Console.ANSI.Codes`. I thought it might be useful to somebody. You'll appreciate that the algorithm is a brute force method and is not optimized for speed. Also, like `xterm256ToSRGB`, it assumes that the system colours are those in `aNSIColors` - which may not be the case for some terminals (or where the terminal allows the system colours to be changed) - such as the new 'campbell' scheme on Windows or the old 'default' Powershell scheme.

Reply:
I see, thanks !